### ISSUE TITLE
Hailo-8: class tensor is pre-sigmoid, confidence clamps to 1.00 on ~30% of detections

### ISSUE THREAD

**Issue by StopItShef** [Keystone-Infrastructure-Corp/nexus-edge-ai-core-next#334]
## Summary

On Hailo-8 cores, `decode_yolo26_raw` consumes the model's class tensor as "post-sigmoid probabilities" (its own doc comment). The shipped `yolo26n_512x288_hailo.hef` does not apply sigmoid on the class branch — it emits raw activations with a dequantization scale of `0.1662890`, giving a value range of roughly `0 .. 42.4`.

Because the tensor is UINT8-quantized, the dequantized scores land on a coarse ladder with a step of 0.166. Everything at step ≥ 7 exceeds 1.0 and is silently absorbed by `score.clamp(0.0, 1.0)` in `hailo_yolo.rs:237`.

Net effect: **confidence takes only ~6 distinct values per output layer, and 28–30% of all detections report exactly `1.000000`.** Rule thresholds such as `object.confidence >= 0.6` cannot discriminate, and the most badly wrong detections receive the highest score.

## Affected scope

| | |
|---|---|
| Hardware | Hailo-8 M.2 (`HLDDMBA234605582`, fw 4.24.0), AMD Radeon 680M host, Ubuntu 24.04 |
| Core | `nals-tx3083herschelave-1`, `core_id f9410902-3684-4cf0-9af3-fd5302682bd6` |
| Engine versions | reproduced on **0.1.211 and 0.1.219** |
| Model | `yolo26n_512x288_hailo.hef` |
| Layout | `RawYolo26 { num_classes: 80, scales: [stride 8 → 64×36, stride 16 → 32×18, stride 32 → 16×9] }` |

Not reproducible on ONNX/CPU/GPU cores — the defect is in the HEF's class branch, and that path is Hailo-only.

### 1. HEF tensor declarations

```
$ hailortcli parse-hef /opt/nexus/current/share/models/yolo26n_512x288_hailo.hef
Architecture HEF was compiled for: HAILO8
Network group name: yolo26n_512x288, Multi Context - Number of contexts: 3
    Network name: yolo26n_512x288/yolo26n_512x288
        VStream infos:
            Input  yolo26n_512x288/input_layer1  UINT8, NHWC(288x512x3)
            Output yolo26n_512x288/conv61        UINT8, NHWC(36x64x4)     <- box, stride 8
            Output yolo26n_512x288/conv77        UINT8, NHWC(18x32x4)     <- box, stride 16
            Output yolo26n_512x288/conv91        UINT8, NHWC(9x16x4)      <- box, stride 32
            Output yolo26n_512x288/conv64        UINT8, FCR(36x64x80)     <- class, stride 8
            Output yolo26n_512x288/conv80        UINT8, FCR(18x32x80)     <- class, stride 16
            Output yolo26n_512x288/conv94        UINT8, FCR(9x16x80)      <- class, stride 32
```

Box branch is `C=4` — DFL **is** projected on-chip, in NHWC, exactly as the decoder contract requires. **The box branch is correct; it is not part of this bug.**

### 2. The dequantization ladder

Every confidence value observed across 300 detections (3 separate diagnostics exports, 2 engine versions, 3 days) is an exact integer multiple of `0.1662890`:

```
0.332578  =  2 x 0.1662890   exact
0.498867  =  3 x 0.1662890   exact
0.665157  =  4 x 0.1662890   exact
0.831446  =  5 x 0.1662890   exact
0.997735  =  6 x 0.1662890   exact
1.164023  =  7 x 0.1662890   --> clamped to 1.000000
1.330312  =  8 x 0.1662890   --> clamped to 1.000000
```

Two further observed values (`0.559244`, `0.861431`) are not on this ladder, which is consistent with the three class layers each carrying their own `qp_scale`. Only one ladder had enough samples to solve.

### 3. The scale cannot belong to a probability tensor

| | |
|---|---|
| Scale needed for UINT8 over `[0,1]` | `1/255 ≈ 0.0039` |
| Actual scale | `0.1662890` |
| Ratio | **43x** |
| Implied tensor range | `0 .. 42.4` |

A post-sigmoid probability tensor has no use for a range of 42.

### 4. Corroborating signal — the thresholds collapse

`SCORE_FLOOR = 0.20` (`imp.rs:938`) and the operator-facing `inference.score_threshold = 0.30` both fall between ladder step 1 (`0.166`) and step 2 (`0.333`). Both therefore reduce to the same predicate, "≥ 2 steps."

This predicts that `0.332578` is the minimum observable confidence on this hardware. It is — in every sample taken, on both engine versions.

## Root cause

`crates/nexus-hailo-backend/src/imp.rs:929` — `decode_yolo26_raw` doc comment:

> *"Box tensor encodes (l, t, r, b) per cell in CELL units (already DFL-projected by the chip). Class tensor encodes per-class **post-sigmoid probabilities**."*

The decoder reads the class value and uses it directly as the score (`imp.rs`, best-class loop ≈ lines 972–987). No sigmoid is applied anywhere in the Hailo path. The HEF does not apply it either.

The overflow is then hidden downstream:

```rust
// crates/nexus-inference/src/hailo_yolo.rs:237
confidence: det.score.clamp(0.0, 1.0),
```

That clamp is a no-op if the contract holds. It is currently absorbing ~30% of all detections.

## Already ruled out — please don't re-investigate

| Hypothesis | Status | Why |
|---|---|---|
| Raw UINT8 being misread as f32 | **Ruled out** | `imp.rs:317` requests `HAILO_FORMAT_TYPE_FLOAT32` output vstreams; HailoRT dequantizes. Input is deliberately UINT8 (`imp.rs:293`). |
| Box DFL bins misread as box edges | **Ruled out** | `hailortcli` shows box outputs at `C=4`, not `C=64`. DFL is projected on-chip. |
| Grid/stride geometry wrong | **Ruled out** | All three scales verified: 64x8=512, 36x8=288; 32x16=512, 18x16=288; 16x32=512, 9x32=288. `input_w`/`input_h` derivation in `decode_yolo26_raw` yields 512x288 correctly. |
| Missing sigmoid relative to the ONNX path | **Not the framing** | `crates/nexus-inference/src/yolo.rs` also applies no sigmoid. Both decoders assume the model bakes it in; exported YOLO ONNX does, this HEF does not. |

### Open, lower priority

Box outputs are declared `NHWC`, class outputs `FCR`. `imp.rs:317–320` forces NHWC on all outputs at vstream creation ("so the raw-YOLO decoder doesn't have to translate FCR"). Empirically class labels are mostly plausible, so the override appears to work — but the asymmetry was not verified and is worth a look while someone is in this code.

## Fix

Two options. **They are mutually exclusive — applying both double-sigmoids.**

**A. Recompile the HEF with sigmoid in the class branch.** Makes the shipped model satisfy the contract the decoder already documents. No engine change. Requires reissuing the HEF to every Hailo core, and any core running an old HEF against a new engine stays broken.

**B. Apply sigmoid in `decode_yolo26_raw` when layout is `RawYolo26`.** One-line change, ships with the engine, fixes existing deployments on upgrade. But it breaks any HEF that *does* bake in sigmoid, so it needs a guard or a guarantee about what all shipped HEFs do.

Whoever owns HEF compilation should pick. If B, `SCORE_FLOOR` (`imp.rs:938`) and the default `inference.score_threshold` need re-checking — both were chosen against a `[0,1]` assumption that has never held on this hardware.

### What the fix does and does not buy

Running the ladder through the sigmoid it never got:

| Raw (steps) | Reported today | True probability |
|---|---|---|
| 2 | 0.33 | 0.58 |
| 4 | 0.67 | 0.66 |
| 6 | 1.00 | 0.73 |
| 10 | 1.00 | 0.84 |
| 20 | 1.00 | 0.96 |

**Fixes:** confidence becomes continuous and comparable; rule thresholds start working; the 76%–99.9% band stops collapsing into a single value.

**Does not fix:** high-confidence misclassifications (furniture as `person`, gym equipment as `vehicle.motorcycle`). Those carry genuinely high logits — sigmoid reports them as 0.98 instead of 1.00. That is a model-capacity question: `yolo26n` (nano) at 512x288 is the only operating point available on Hailo, against a 640x640 default elsewhere. Worth a separate issue.

## Reproduction

```bash
# 1. Confirm the HEF's class branch is FCR/UINT8 at C=80 (read-only, safe on a live core)
hailortcli parse-hef /opt/nexus/current/share/models/yolo26n_512x288_hailo.hef

# 2. Pull diagnostics and inspect the confidence distribution.
#    Expect: <10 distinct values across 100 detections; ~30% at exactly 1.0;
#    minimum exactly 0.332578
python3 -c "
import json,collections
d=json.load(open('motion-events.json'))
c=collections.Counter(round(e['confidence'],6) for e in d)
print(sorted(c.items()))
print('exactly 1.0:', sum(n for v,n in c.items() if v>=0.9999), '/', len(d))
"

# 3. Verify the ladder
python3 -c "
step=0.1662890
for v in [0.332578,0.498867,0.665157,0.831446,0.997735]:
    print(v, v/step)
"
```

`nexus-hailo-probe` would report the same tensor info but is **not shipped** in `/opt/nexus/current/bin/` (only `nexus-engine`, `nexus-doctor`, `nexus-probe`). Worth adding to the install payload — `hailortcli` covered it here, but that is a HailoRT dependency rather than ours.

## Key code references

| Path | Line | What |
|---|---|---|
| `crates/nexus-hailo-backend/src/imp.rs` | 929 | `decode_yolo26_raw` — the contract comment |
| `crates/nexus-hailo-backend/src/imp.rs` | 938 | `SCORE_FLOOR = 0.20` |
| `crates/nexus-hailo-backend/src/imp.rs` | 972–987 | best-class loop; raw value used as score |
| `crates/nexus-hailo-backend/src/imp.rs` | 293 | input vstream forced UINT8 |
| `crates/nexus-hailo-backend/src/imp.rs` | 317 | output vstreams forced FLOAT32 + NHWC |
| `crates/nexus-inference/src/hailo_yolo.rs` | 237 | `score.clamp(0.0, 1.0)` — masks the overflow |
| `crates/nexus-inference/src/yolo.rs` | ~367–400 | ONNX decode, for comparison |

**Comment by StopItShef** [comment]
Related: #335 — tracker frame-count thresholds are calibrated for 30fps and the static-object filter cannot promote on Hailo-8 hardware. Separate root cause and separate fix, but the two compound in production: the pre-sigmoid scoring described here produces phantom stationary `person` detections, and #335 covers why nothing suppresses them.
